File: apps/android/phrenkit/src/main/kotlin/com/phren/kit/FindingsFile.kt
package com.phren.kit

import java.security.SecureRandom
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

/**
 * Parser + mutator for a project's FINDINGS.md (port of FindingsFile.swift).
 *
 * Parsing transcribes `readFindings` (packages/cli/src/data/access.ts:276);
 * mutations transcribe `addFindingToFile` (content/learning.ts:298),
 * `editFinding` (access.ts:558), and `removeFinding` (access.ts:462).
 */
class FindingsFile(content: String) {
    var content: String = content
        private set

    // Parse (access.ts:276 readFindings)

    fun parse(includeArchived: Boolean = false): List<Finding> {
        if (content.isEmpty()) return emptyList()
        val lines = content.split("\n")
        val items = mutableListOf<Finding>()
        var date = "unknown"
        var index = 1
        var inArchiveBlock = false
        var headingTag: String? = null

        var i = 0
        while (i < lines.size) {
            val line = lines[i]
            try {
                if (isArchiveStart(line)) { inArchiveBlock = true; continue }
                if (isArchiveEnd(line)) { inArchiveBlock = false; continue }
                if (inArchiveBlock && !includeArchived) continue

                extractDateHeading(line)?.let { date = it; continue }

                // access.ts:315 — heading-based findings: ## topic / ### title / paragraph
                val h2Tag = H2_TAG.group(line)
                if (h2Tag != null && !H2_YEAR.test(line)) {
                    headingTag = h2Tag.lowercase()
                    continue
                }
                val h3Title = H3.group(line)
                val tag = headingTag
                if (h3Title != null && tag != null) {
                    var body = ""
                    var j = i + 1
                    while (j < lines.size) {
                        val next = lines[j].jsTrimmed
                        j++
                        if (next.isEmpty()) continue
                        if (next.startsWith("#") || next.startsWith("- ")) break
                        body = next
                        break
                    }
                    val title = h3Title.jsTrimmed
                    val syntheticText = if (body.isEmpty()) "[$tag] $title" else "[$tag] $title — $body"
                    items += Finding(
                        id = "L$index", date = date, text = syntheticText,
                        archived = inArchiveBlock, typeTag = tag, rawLine = line,
                    )
                    index++
                    continue
                }

                if (!line.startsWith("- ")) continue

                val next = if (i + 1 < lines.size) lines[i + 1] else ""
                val citation = if (isCitationLine(next)) next.jsTrimmed else null
                val citationData = citation?.let(::parseCitationComment)
                val provenance = parseSourceComment(line)
                val scope = parseScopeComment(line) ?: provenance?.scope
                val stableId = parseFindingId(line)
                val rawText = LEADING_BULLET.replaceFirst(line, "").jsTrimmed
                val textWithoutComments = stripComments(rawText)
                var confidence: Double? = null
                var text = textWithoutComments
                CONFIDENCE_SUFFIX.firstMatch(textWithoutComments)?.let { m ->
                    confidence = m.group(1).toDoubleOrNull()
                    text = textWithoutComments.substring(0, m.start()).jsTrimmed
                }

                val contradictsRefs = parseAllContradictions(line)
                val lifecycle = parseFindingLifecycle(line)
                items += Finding(
                    id = "L$index",
                    stableId = stableId,
                    date = date,
                    text = text,
                    citation = citation,
                    citationData = citationData,
                    taskItem = citationData?.taskItem,
                    confidence = confidence,
                    scope = scope,
                    machine = provenance?.machine,
                    actor = provenance?.actor,
                    supersededBy = MetadataRegex.supersededBy.group(line),
                    supersedes = MetadataRegex.supersedes.group(line),
                    contradicts = contradictsRefs.ifEmpty { null },
                    status = lifecycle.status,
                    statusUpdated = lifecycle.statusUpdated,
                    statusReason = lifecycle.statusReason,
                    statusRef = lifecycle.statusRef,
                    archived = inArchiveBlock,
                    typeTag = TYPE_PREFIX.group(text)?.lowercase(),
                    rawLine = line,
                )
                if (citation != null) i++
                index++
            } finally {
                i++
            }
        }
        return items
    }

    /**
     * The date of the last consolidation, from the `<!-- consolidated: … -->`
     * stamp `autoArchiveToReference` writes (content/archive.ts:236).
     */
    val consolidatedDate: String? get() = CONSOLIDATED.group(content)

    // Add (learning.ts:160 prepareFinding + 244 insertFindingIntoContent)

    data class AddOptions(
        val type: FindingType? = null,
        val scope: String? = null,
        val provenance: FindingProvenance? = null,
        /** Set when promoting a review-queue item: the date it was queued. */
        val queuedDate: String? = null,
        val now: Instant = Instant.now(),
    )

    /**
     * Adds a finding, mirroring the CLI's bullet construction. Intentional MVP
     * divergences (apps/ios/README.md): no coref, no auto type detection,
     * exact-normalized dedup only, no auto-archive cap.
     */
    fun add(project: String, text: String, options: AddOptions = AddOptions()): String {
        val learning = text.jsTrimmed
        if (learning.isEmpty()) throw PhrenKitError.EmptyInput("Finding text cannot be empty.")
        SecretScanner.scan(learning)?.let {
            throw PhrenKitError.SecretDetected("Rejected: finding appears to contain a secret ($it). Strip credentials before saving.")
        }
        val nowIso = isoTimestamp(options.now)
        val today = nowIso.take(10)

        var normalizedLearning = learning
        // core/finding.ts `applyFindingTypePrefix`: the test is anchored and accepts any
        // bracketed tag. `extractFindingType` is unanchored and knows only the decay types,
        // so a `[bug]` mid-sentence dropped the caller's type and `[tradeoff]` was tagged twice.
        if (options.type != null && !FINDING_TAG_PREFIX.test(normalizedLearning)) {
            normalizedLearning = "[${options.type.rawValue}] $normalizedLearning"
        }

        val fid = randomHexId()
        var bullet = if (normalizedLearning.startsWith("- ")) normalizedLearning else "- $normalizedLearning"
        bullet += " <!-- fid:$fid --> <!-- created: $today -->"
        val scopeComment = buildScopeComment(options.scope)
        if (scopeComment.isNotEmpty()) bullet += " $scopeComment"
        options.provenance?.let {
            val sourceComment = buildSourceComment(it)
            if (sourceComment.isNotEmpty()) bullet += " $sourceComment"
        }
        // A promoted queue item is written today but was captured earlier: learning.ts
        // places approveQueueItemDetailed's extra annotation after scope and source.
        if (!options.queuedDate.isNullOrEmpty()) bullet += " <!-- phren:queued \"${options.queuedDate}\" -->"

        if (isDuplicate(bullet)) {
            throw PhrenKitError.Duplicate("Skipped duplicate finding for \"$project\": already exists with similar wording.")
        }

        bullet += " " + buildLifecycleComments(
            FindingLifecycleMetadata(status = FindingLifecycleStatus.ACTIVE, statusUpdated = today),
            fallbackDate = today,
        )
        val citationComment = "  " + buildCitationComment(FindingCitation(createdAt = nowIso))

        content = if (content.isEmpty()) {
            // learning.ts:349 — brand-new FINDINGS.md
            "# $project Findings\n\n## $today\n\n$bullet\n$citationComment\n"
        } else {
            insertFindingIntoContent(content, today, bullet, citationComment)
        }
        return fid
    }

    /** Exact-normalized-text duplicate check — minimal port of `isDuplicateFinding` (dedup.ts:394). */
    private fun isDuplicate(bullet: String): Boolean {
        val needle = normalizeFindingText(bullet)
        if (needle.isEmpty()) return false
        return content.split("\n").filter { it.startsWith("- ") }.any { normalizeFindingText(it) == needle }
    }

    // Edit (access.ts:558 editFinding)

    fun edit(project: String, oldText: String, newText: String) {
        val newTextTrimmed = newText.jsTrimmed
        if (newTextTrimmed.isEmpty()) throw PhrenKitError.EmptyInput("New finding text cannot be empty.")
        val lines = content.split("\n").toMutableList()
        val idx = matchBullet(lines, oldText, project)

        // access.ts:588 — preserve every metadata comment as a re-appended
        // suffix, and keep the [tag] prefix unless the new text supplies one.
        val existing = lines[idx]
        val metaComments = MetadataRegex.anyComment.allMatches(existing)
        val metaSuffix = if (metaComments.isEmpty()) "" else " " + metaComments.joinToString(" ")
        val existingTag = EXISTING_TAG.group(existing)
        val newHasTag = NEW_TAG.test(newTextTrimmed)
        val tagPrefix = if (existingTag != null && !newHasTag) "$existingTag " else ""
        lines[idx] = "- $tagPrefix$newTextTrimmed$metaSuffix"
        content = normalizeWrite(lines)
    }

    // Remove (access.ts:462 removeFinding)

    fun remove(project: String, match: String): String {
        val lines = content.split("\n").toMutableList()
        val idx = matchBullet(lines, match, project)
        val removeCount = if (idx + 1 < lines.size && isCitationLine(lines[idx + 1])) 2 else 1
        val matched = lines[idx]
        repeat(removeCount) { lines.removeAt(idx) }
        content = normalizeWrite(lines)
        return matched
    }

    // Matching (access.ts:184-248)

    private data class BulletLine(val line: String, val i: Int, val archived: Boolean)

    private fun collectBulletLines(lines: List<String>): List<BulletLine> {
        val bullets = mutableListOf<BulletLine>()
        var inArchiveBlock = false
        lines.forEachIndexed { i, line ->
            if (isArchiveStart(line)) { inArchiveBlock = true; return@forEachIndexed }
            if (isArchiveEnd(line)) { inArchiveBlock = false; return@forEachIndexed }
            if (!line.startsWith("- ")) return@forEachIndexed
            bullets += BulletLine(line, i, inArchiveBlock)
        }
        return bullets
    }

    /** access.ts:204 `bulletContentKey` */
    private fun bulletContentKey(line: String): String =
        MetadataRegex.anyComment.replaceAll(line, " ").collapsedWhitespace.jsTrimmed

    /** access.ts:214 `resolveDuplicateMatches` */
    private fun resolveDuplicateMatches(matches: List<BulletLine>): BulletLine? {
        val first = matches.firstOrNull() ?: return null
        val key = bulletContentKey(first.line)
        return if (matches.all { bulletContentKey(it.line) == key }) first else null
    }

    private fun matchBullet(lines: List<String>, match: String, project: String): Int {
        val needle = normalizeFindingText(match)
        val bullets = collectBulletLines(lines)
        return when (val r = matchIn(bullets.filter { !it.archived }, needle, match)) {
            is MatchResult.Found -> r.index
            is MatchResult.Ambiguous -> throw PhrenKitError.AmbiguousMatch(r.error)
            MatchResult.NotFound -> when (matchIn(bullets.filter { it.archived }, needle, match)) {
                MatchResult.NotFound -> throw PhrenKitError.NotFound("No finding matching \"$match\" in project \"$project\".")
                else -> throw PhrenKitError.ArchivedReadOnly(
                    "Finding \"$match\" is archived and read-only. Restore or re-add it before mutating history.",
                )
            }
        }
    }

    private sealed interface MatchResult {
        data class Found(val index: Int) : MatchResult
        data class Ambiguous(val error: String) : MatchResult
        data object NotFound : MatchResult
    }

    /**
     * access.ts `existsAsLiveFinding`: does this text already exist as a live
     * (non-archived) bullet? Ambiguous counts as present: approve must not write another copy.
     */
    fun existsAsLiveFinding(text: String): Boolean {
        val needle = normalizeFindingText(text)
        if (needle.isEmpty()) return false
        val active = collectBulletLines(content.split("\n")).filter { !it.archived }
        return matchIn(active, needle, text) != MatchResult.NotFound
    }

    private fun matchIn(bullets: List<BulletLine>, needle: String, match: String): MatchResult {
        val fidNeedle = if (needle.startsWith("fid:")) needle.drop(4) else needle
        if (FID_NEEDLE.test(fidNeedle)) {
            val fidRegex = JSRegex("<!--\\s*fid:$fidNeedle\\s*-->")
            val fidMatches = bullets.filter { fidRegex.test(it.line) }
            if (fidMatches.size == 1) return MatchResult.Found(fidMatches[0].i)
        }
        val exact = bullets.filter { normalizeFindingText(it.line) == needle }
        if (exact.size == 1) return MatchResult.Found(exact[0].i)
        if (exact.size > 1) {
            resolveDuplicateMatches(exact)?.let { return MatchResult.Found(it.i) }
            return MatchResult.Ambiguous("\"$match\" is ambiguous (${exact.size} exact matches). Use a more specific phrase.")
        }
        val partial = bullets.filter { normalizeFindingText(it.line).contains(needle) }
        if (partial.size == 1) return MatchResult.Found(partial[0].i)
        if (partial.size > 1) {
            resolveDuplicateMatches(partial)?.let { return MatchResult.Found(it.i) }
            return MatchResult.Ambiguous("\"$match\" is ambiguous (${partial.size} partial matches). Use a more specific phrase.")
        }
        return MatchResult.NotFound
    }

    companion object {
        /** core/finding.ts `FINDING_TAG_PREFIX_RE` */
        internal val FINDING_TAG_PREFIX = JSRegex("""^\s*\[[^\]]+\]\s*""")

        private val H2_TAG = JSRegex("""^##\s+([a-z_-]+)\s*$""", caseInsensitive = true)
        private val H2_YEAR = JSRegex("""^##\s+\d{4}""")
        private val H3 = JSRegex("""^###\s+(.+)$""")
        private val LEADING_BULLET = JSRegex("""^-\s+""")
        private val CONFIDENCE_SUFFIX = JSRegex("""\s*\[confidence\s+([01](?:\.\d+)?)\]\s*$""", caseInsensitive = true)
        private val TYPE_PREFIX = JSRegex("""^\[([a-z][a-z0-9_-]*)\]""", caseInsensitive = true)
        private val CONSOLIDATED = JSRegex("""<!--\s*consolidated:\s*(\d{4}-\d{2}-\d{2})""")
        private val DATE_HEADING = JSRegex("""^##\s+(.+)$""")
        private val DATE_ONLY = JSRegex("""^(\d{4}-\d{2}-\d{2})$""")
        private val ARCHIVED_DATE = JSRegex("""^Archived\s+(\d{4}-\d{2}-\d{2})$""", caseInsensitive = true)
        private val FIRST_DATE_HEADING = JSRegex.multiline("""^## \d{4}-\d{2}-\d{2}""")
        private val EXISTING_TAG = JSRegex("""^-\s*(\[[a-z][a-z0-9_-]*\])\s""")
        private val NEW_TAG = JSRegex("""^\[[a-z][a-z0-9_-]*\]""")
        private val FID_NEEDLE = JSRegex("""^[a-z0-9]{8}$""")
        private val TRIPLE_NEWLINES = JSRegex("""\n{3,}""")
        private val random = SecureRandom()
        private val ISO = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").withZone(ZoneOffset.UTC)

        fun empty(@Suppress("UNUSED_PARAMETER") project: String) = FindingsFile("")

        /** access.ts:164 `extractDateHeading` */
        fun extractDateHeading(line: String): String? {
            val raw = DATE_HEADING.group(line)?.jsTrimmed ?: return null
            DATE_ONLY.group(raw)?.let { return it }
            return ARCHIVED_DATE.group(raw)
        }

        /**
         * learning.ts:244 `insertFindingIntoContent` — the today-header search
         * starts after the last `</details>` so archived blocks are never reused.
         */
        fun insertFindingIntoContent(content: String, today: String, bullet: String, citationComment: String): String {
            val todayHeader = "## $today"
            val searchFrom = content.lastIndexOf("</details>").coerceAtLeast(0)
            val header = content.indexOf(todayHeader, searchFrom)
            if (header >= 0) {
                val insertAt = header + todayHeader.length
                return content.substring(0, insertAt) + "\n\n$bullet\n$citationComment" + content.substring(insertAt)
            }
            FIRST_DATE_HEADING.firstMatch(content)?.let { m ->
                return content.substring(0, m.start()) + "$todayHeader\n\n$bullet\n$citationComment\n\n" + content.substring(m.start())
            }
            return trimEnd(content) + "\n\n## $today\n\n$bullet\n$citationComment\n"
        }

        /** `lines.join("\n").replace(/\n{3,}/g, "\n\n").trimEnd() + "\n"` (access.ts:493). */
        fun normalizeWrite(lines: List<String>): String = normalizeContent(lines.joinToString("\n"))

        fun normalizeContent(joined: String): String = trimEnd(TRIPLE_NEWLINES.replaceAll(joined, "\n\n")) + "\n"

        /** crypto.randomBytes(4).toString("hex") */
        fun randomHexId(): String {
            val bytes = ByteArray(4).also(random::nextBytes)
            return bytes.joinToString("") { "%02x".format(it.toInt() and 0xFF) }
        }

        /** new Date().toISOString() */
        fun isoTimestamp(date: Instant): String = ISO.format(date)
    }
}

/** JS `String.prototype.trimEnd()`. */
fun trimEnd(s: String): String = s.trimEnd(JSRegex::isJsWhitespace)

File: apps/ios/PhrenKit/Sources/PhrenKit/Markdown/FindingsFile.swift
import Foundation

/// Parser + mutator for a project's FINDINGS.md.
///
/// Parsing transcribes `readFindings` (packages/cli/src/data/access.ts:276);
/// mutations transcribe `addFindingToFile` (content/learning.ts:298),
/// `editFinding` (access.ts:558), and `removeFinding` (access.ts:462).
/// Mutations are surgical line edits followed by the CLI's exact
/// normalization pass — nothing else is rewritten.
public struct FindingsFile: Sendable {
    public private(set) var content: String

    public init(content: String) {
        self.content = content
    }

    public static func empty(project: String) -> FindingsFile {
        FindingsFile(content: "")
    }

    // MARK: - Parse (access.ts:276 readFindings)

    public func parse(includeArchived: Bool = false) -> [Finding] {
        guard !content.isEmpty else { return [] }
        let lines = content.components(separatedBy: "\n")
        var items: [Finding] = []
        var date = "unknown"
        var index = 1
        var inArchiveBlock = false
        var headingTag: String?

        var i = 0
        while i < lines.count {
            defer { i += 1 }
            let line = lines[i]
            if isArchiveStart(line) { inArchiveBlock = true; continue }
            if isArchiveEnd(line) { inArchiveBlock = false; continue }
            if inArchiveBlock && !includeArchived { continue }

            if let extracted = Self.extractDateHeading(line) {
                date = extracted
                continue
            }

            // access.ts:315 — heading-based findings: ## topic / ### title / paragraph
            if let h2Tag = JSRegex(#"^##\s+([a-z_-]+)\s*$"#, caseInsensitive: true).group(line),
               !JSRegex(#"^##\s+\d{4}"#).test(line) {
                headingTag = h2Tag.lowercased()
                continue
            }
            if let h3Title = JSRegex(#"^###\s+(.+)$"#).group(line), let tag = headingTag {
                var body = ""
                var j = i + 1
                while j < lines.count {
                    let next = lines[j].jsTrimmed
                    j += 1
                    if next.isEmpty { continue }
                    if next.hasPrefix("#") || next.hasPrefix("- ") { break }
                    body = next
                    break
                }
                let title = h3Title.jsTrimmed
                let syntheticText = body.isEmpty ? "[\(tag)] \(title)" : "[\(tag)] \(title) — \(body)"
                items.append(Finding(
                    id: "L\(index)", stableId: nil, date: date, text: syntheticText,
                    citation: nil, citationData: nil, taskItem: nil, confidence: nil,
                    scope: nil, machine: nil, actor: nil, supersededBy: nil,
                    supersedes: nil, contradicts: nil, status: .active,
                    statusUpdated: nil, statusReason: nil, statusRef: nil,
                    archived: inArchiveBlock, typeTag: tag, rawLine: line,
                    journalFile: nil
                ))
                index += 1
                continue
            }

            guard line.hasPrefix("- ") else { continue }

            let next = i + 1 < lines.count ? lines[i + 1] : ""
            let citation = isCitationLine(next) ? next.jsTrimmed : nil
            let citationData = citation.flatMap(parseCitationComment)
            let provenance = parseSourceComment(line)
            let scope = parseScopeComment(line) ?? provenance?.scope
            let stableId = parseFindingId(line)
            let rawText = JSRegex(#"^-\s+"#).replaceFirst(line, with: "").jsTrimmed
            let textWithoutComments = stripComments(rawText)
            let confRegex = JSRegex(#"\s*\[confidence\s+([01](?:\.\d+)?)\]\s*$"#, caseInsensitive: true)
            var confidence: Double?
            var text = textWithoutComments
            if let m = confRegex.firstMatch(in: textWithoutComments),
               let whole = JSRegex.substring(textWithoutComments, m, 0),
               let value = JSRegex.substring(textWithoutComments, m, 1) {
                confidence = Double(value)
                text = String(textWithoutComments.dropLast(whole.count)).jsTrimmed
            }

            let supersededByRef = MetadataRegex.supersededBy.group(line)
            let supersedesRef = MetadataRegex.supersedes.group(line)
            let contradictsRefs = parseAllContradictions(line)
            let lifecycle = parseFindingLifecycle(line)
            let typeTag = JSRegex(#"^\[([a-z][a-z0-9_-]*)\]"#, caseInsensitive: true).group(text)?.lowercased()

            items.append(Finding(
                id: "L\(index)",
                stableId: stableId,
                date: date,
                text: text,
                citation: citation,
                citationData: citationData,
                taskItem: citationData?.taskItem,
                confidence: confidence,
                scope: scope,
                machine: provenance?.machine,
                actor: provenance?.actor,
                supersededBy: supersededByRef,
                supersedes: supersedesRef,
                contradicts: contradictsRefs.isEmpty ? nil : contradictsRefs,
                status: lifecycle.status,
                statusUpdated: lifecycle.statusUpdated,
                statusReason: lifecycle.statusReason,
                statusRef: lifecycle.statusRef,
                archived: inArchiveBlock,
                typeTag: typeTag,
                rawLine: line,
                journalFile: nil
            ))
            if citation != nil { i += 1 }
            index += 1
        }
        return items
    }

    /// The date of the last consolidation, from the `<!-- consolidated: … -->`
    /// stamp `autoArchiveToReference` writes into the file it just emptied
    /// (content/archive.ts:236). Matching is the CLI's own
    /// (content/validate.ts:56) — leading whitespace tolerated, trailing text
    /// ignored — so a marker either side reads the same date.
    ///
    /// This is what lets the app tell "this project has never been
    /// consolidated" from "everything older than N findings moved to the cold
    /// tier", using a file it already syncs and at no extra cost.
    public var consolidatedDate: String? {
        JSRegex(#"<!--\s*consolidated:\s*(\d{4}-\d{2}-\d{2})"#).group(content)
    }

    /// access.ts:164 `extractDateHeading`
    static func extractDateHeading(_ line: String) -> String? {
        guard let raw = JSRegex(#"^##\s+(.+)$"#).group(line)?.jsTrimmed else { return nil }
        if let direct = JSRegex(#"^(\d{4}-\d{2}-\d{2})$"#).group(raw) { return direct }
        if let archived = JSRegex(#"^Archived\s+(\d{4}-\d{2}-\d{2})$"#, caseInsensitive: true).group(raw) {
            return archived
        }
        return nil
    }

    // MARK: - Add (learning.ts:160 prepareFinding + 244 insertFindingIntoContent)

    public struct AddOptions: Sendable {
        public var type: FindingType?
        public var scope: String?
        public var provenance: FindingProvenance?
        public var now: Date

        public init(type: FindingType? = nil, scope: String? = nil,
                    provenance: FindingProvenance? = nil, now: Date = Date()) {
            self.type = type
            self.scope = scope
            self.provenance = provenance
            self.now = now
        }
    }

    /// Adds a finding, mirroring the CLI's bullet construction. Intentional MVP
    /// divergences from `addFindingToFile` (documented in apps/ios/README.md):
    /// no coref resolution, no auto type detection, no semantic (Jaccard)
    /// dedup — only exact normalized-text dedup — and no auto-archive cap.
    /// The CLI tidies all of these on its next run.
    @discardableResult
    public mutating func add(project: String, text: String, options: AddOptions = AddOptions()) throws -> String {
        let learning = text.jsTrimmed
        guard !learning.isEmpty else { throw PhrenKitError.emptyInput("Finding text cannot be empty.") }
        if let secret = SecretScanner.scan(learning) {
            throw PhrenKitError.secretDetected(
                "Rejected: finding appears to contain a secret (\(secret)). Strip credentials before saving."
            )
        }

        let nowIso = Self.isoTimestamp(options.now)
        let today = String(nowIso.prefix(10))

        var normalizedLearning = learning
        // core/finding.ts:16-28 `applyFindingTypePrefix`: the test is anchored
        // and accepts any bracketed tag. `extractFindingType` is unanchored, so a
        // `[bug]` mid-sentence suppressed the caller's type, and it knows only the
        // decay types, so `[tradeoff]`/`[architecture]` were tagged twice.
        if let type = options.type, !Self.findingTagPrefix.test(normalizedLearning) {
            normalizedLearning = "[\(type.rawValue)] \(normalizedLearning)"
        }

        let fid = Self.randomHexId()
        var bullet = normalizedLearning.hasPrefix("- ") ? normalizedLearning : "- \(normalizedLearning)"
        bullet += " <!-- fid:\(fid) --> <!-- created: \(today) -->"
        let scopeComment = buildScopeComment(options.scope)
        if !scopeComment.isEmpty { bullet += " \(scopeComment)" }
        if let provenance = options.provenance {
            let sourceComment = buildSourceComment(provenance)
            if !sourceComment.isEmpty { bullet += " \(sourceComment)" }
        }

        if isDuplicate(of: bullet) {
            throw PhrenKitError.duplicate(
                "Skipped duplicate finding for \"\(project)\": already exists with similar wording."
            )
        }

        bullet += " " + buildLifecycleComments(
            FindingLifecycleMetadata(status: .active, statusUpdated: today),
            fallbackDate: today
        )
        let citationComment = "  " + buildCitationComment(FindingCitation(createdAt: nowIso))

        if content.isEmpty {
            // learning.ts:349 — brand-new FINDINGS.md
            content = "# \(project) Findings\n\n## \(today)\n\n\(bullet)\n\(citationComment)\n"
        } else {
            content = Self.insertFindingIntoContent(content, today: today, bullet: bullet, citationComment: citationComment)
        }
        return fid
    }

    /// Exact-normalized-text duplicate check against existing bullets — the
    /// minimal port of `isDuplicateFinding` (dedup.ts:394).
    private func isDuplicate(of bullet: String) -> Bool {
        let needle = normalizeFindingText(bullet)
        guard !needle.isEmpty else { return false }
        return content.components(separatedBy: "\n")
            .filter { $0.hasPrefix("- ") }
            .contains { normalizeFindingText($0) == needle }
    }

    /// learning.ts:244 `insertFindingIntoContent` — positional insertion; the
    /// today-header search starts after the last `</details>` so archived
    /// blocks with a matching date are never reused.
    static func insertFindingIntoContent(_ content: String, today: String, bullet: String, citationComment: String) -> String {
        let todayHeader = "## \(today)"
        let searchFrom: String.Index
        if let lastDetailsClose = content.range(of: "</details>", options: .backwards) {
            searchFrom = lastDetailsClose.lowerBound
        } else {
            searchFrom = content.startIndex
        }
        if let headerRange = content.range(of: todayHeader, range: searchFrom..<content.endIndex) {
            let insertAt = headerRange.upperBound
            return String(content[..<insertAt]) + "\n\n\(bullet)\n\(citationComment)" + String(content[insertAt...])
        }
        if let m = JSRegex(multiline: #"^## \d{4}-\d{2}-\d{2}"#).firstMatch(in: content),
           let firstHeading = Range(m.range, in: content) {
            return String(content[..<firstHeading.lowerBound])
                + "\(todayHeader)\n\n\(bullet)\n\(citationComment)\n\n"
                + String(content[firstHeading.lowerBound...])
        }
        return trimEnd(content) + "\n\n## \(today)\n\n\(bullet)\n\(citationComment)\n"
    }

    // MARK: - Edit (access.ts:558 editFinding)

    public mutating func edit(project: String, oldText: String, newText: String) throws {
        let newTextTrimmed = newText.jsTrimmed
        guard !newTextTrimmed.isEmpty else { throw PhrenKitError.emptyInput("New finding text cannot be empty.") }

        var lines = content.components(separatedBy: "\n")
        let idx = try matchBullet(lines: lines, match: oldText, project: project)

        // access.ts:588 — preserve every metadata comment as a re-appended
        // suffix, and keep the [tag] prefix unless the new text supplies one.
        let existing = lines[idx]
        let metaComments = MetadataRegex.anyComment.allMatches(existing)
        let metaSuffix = metaComments.isEmpty ? "" : " " + metaComments.joined(separator: " ")
        let existingTag = JSRegex(#"^-\s*(\[[a-z][a-z0-9_-]*\])\s"#).group(existing)
        let newHasTag = JSRegex(#"^\[[a-z][a-z0-9_-]*\]"#).test(newTextTrimmed)
        let tagPrefix = (existingTag != nil && !newHasTag) ? "\(existingTag!) " : ""
        lines[idx] = "- \(tagPrefix)\(newTextTrimmed)\(metaSuffix)"
        content = Self.normalizeWrite(lines)
    }

    // MARK: - Remove (access.ts:462 removeFinding)

    @discardableResult
    public mutating func remove(project: String, match: String) throws -> String {
        var lines = content.components(separatedBy: "\n")
        let idx = try matchBullet(lines: lines, match: match, project: project)

        let removeCount = (idx + 1 < lines.count && isCitationLine(lines[idx + 1])) ? 2 : 1
        let matched = lines[idx]
        lines.removeSubrange(idx..<(idx + removeCount))
        content = Self.normalizeWrite(lines)
        return matched
    }

    // MARK: - Matching (access.ts:184-248)

    private struct BulletLine {
        let line: String
        let i: Int
        let archived: Bool
    }

    private func collectBulletLines(_ lines: [String]) -> [BulletLine] {
        var bullets: [BulletLine] = []
        var inArchiveBlock = false
        for (i, line) in lines.enumerated() {
            if isArchiveStart(line) { inArchiveBlock = true; continue }
            if isArchiveEnd(line) { inArchiveBlock = false; continue }
            guard line.hasPrefix("- ") else { continue }
            bullets.append(BulletLine(line: line, i: i, archived: inArchiveBlock))
        }
        return bullets
    }

    /// access.ts:204 `bulletContentKey`
    private func bulletContentKey(_ line: String) -> String {
        MetadataRegex.anyComment.replaceAll(line, with: " ").collapsedWhitespace.jsTrimmed
    }

    /// access.ts:214 `resolveDuplicateMatches`
    private func resolveDuplicateMatches(_ matches: [BulletLine]) -> BulletLine? {
        guard let first = matches.first else { return nil }
        let key = bulletContentKey(first.line)
        return matches.allSatisfy { bulletContentKey($0.line) == key } ? first : nil
    }

    /// access.ts:219 `findMatchingFindingBullet` + the archived-check wrapper
    /// shared by editFinding/removeFinding.
    private func matchBullet(lines: [String], match: String, project: String) throws -> Int {
        let needle = normalizeFindingText(match)
        let bullets = collectBulletLines(lines)
        let active = bullets.filter { !$0.archived }
        let archived = bullets.filter { $0.archived }

        switch matchIn(active, needle: needle, match: match) {
        case .found(let idx):
            return idx
        case .ambiguous(let error):
            throw PhrenKitError.ambiguousMatch(error)
        case .notFound:
            switch matchIn(archived, needle: needle, match: match) {
            case .found, .ambiguous:
                throw PhrenKitError.archivedReadOnly(
                    "Finding \"\(match)\" is archived and read-only. Restore or re-add it before mutating history."
                )
            case .notFound:
                throw PhrenKitError.notFound("No finding matching \"\(match)\" in project \"\(project)\".")
            }
        }
    }

    private enum MatchResult {
        case found(Int)
        case ambiguous(String)
        case notFound
    }

    private func matchIn(_ bullets: [BulletLine], needle: String, match: String) -> MatchResult {
        let fidNeedle = needle.hasPrefix("fid:") ? String(needle.dropFirst(4)) : needle
        if JSRegex(#"^[a-z0-9]{8}$"#).test(fidNeedle) {
            let fidRegex = JSRegex("<!--\\s*fid:\(fidNeedle)\\s*-->")
            let fidMatches = bullets.filter { fidRegex.test($0.line) }
            if fidMatches.count == 1 { return .found(fidMatches[0].i) }
        }

        let exact = bullets.filter { normalizeFindingText($0.line) == needle }
        if exact.count == 1 { return .found(exact[0].i) }
        if exact.count > 1 {
            if let dup = resolveDuplicateMatches(exact) { return .found(dup.i) }
            return .ambiguous("\"\(match)\" is ambiguous (\(exact.count) exact matches). Use a more specific phrase.")
        }

        let partial = bullets.filter { normalizeFindingText($0.line).contains(needle) }
        if partial.count == 1 { return .found(partial[0].i) }
        if partial.count > 1 {
            if let dup = resolveDuplicateMatches(partial) { return .found(dup.i) }
            return .ambiguous("\"\(match)\" is ambiguous (\(partial.count) partial matches). Use a more specific phrase.")
        }
        return .notFound
    }

    // MARK: - Shared write normalization

    /// The CLI's exact post-write pass (access.ts:493 and equivalents):
    /// `lines.join("\n").replace(/\n{3,}/g, "\n\n").trimEnd() + "\n"`.
    static func normalizeWrite(_ lines: [String]) -> String {
        normalizeContent(lines.joined(separator: "\n"))
    }

    static func normalizeContent(_ joined: String) -> String {
        trimEnd(JSRegex(#"\n{3,}"#).replaceAll(joined, with: "\n\n")) + "\n"
    }

    /// core/finding.ts:16 `FINDING_TAG_PREFIX_RE`.
    static let findingTagPrefix = JSRegex(#"^\s*\[[^\]]+\]\s*"#)

    static func randomHexId() -> String {
        // crypto.randomBytes(4).toString("hex")
        (0..<4).map { _ in String(format: "%02x", UInt8.random(in: 0...255)) }.joined()
    }

    static func isoTimestamp(_ date: Date) -> String {
        // new Date().toISOString(): yyyy-MM-dd'T'HH:mm:ss.SSS'Z'
        PhrenDateFormats.utc("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").string(from: date)
    }
}

/// JS `String.prototype.trimEnd()`.
func trimEnd(_ s: String) -> String {
    var view = Substring(s)
    while let last = view.last, last.isWhitespace || last.isNewline {
        view = view.dropLast()
    }
    return String(view)
}
